NODE_METADATA_DESIGN.md

```diff
@@ -0,0 +1,49 @@
+## Design proposal for additional node metadata - [JENKINS-26522](https://issues.jenkins-ci.org/browse/JENKINS-26522)
+
+### Problem statement
+
+Before this change, `FlowNode`s only can have a limited amount of metadata. For example, status is simply determined by
+whether there is an `ErrorAction` on the `FlowNode` - if so, it's considered failed, otherwise, it's considered passed.
+There's no way to record or report on more granular status, such as `UNSTABLE` or `ABORTED` for an individual `FlowNode`
+(or block/stage). There also is no way to allow plugins to contribute additional metadata to `FlowNode`s, such as test
+results, so there's no way to report on which particular step or stage contributed which test results. Combined, this
+limits the reporting and visualization capabilities available to Blue Ocean and the Pipeline Stage View significantly.
+
+### Summary of Design
+
+To address this, we will use a new `PersistentAction` similar to `TagsAction` to record that needed additional metadata.
+This new action, named something like `PropagatingTagsAction` to signify that it plays a similar role to `TagsAction`, 
+with the noteworthy difference that its tags are propagated to enclosing blocks. Status in particular will be 
+special-cased in `PropagatingTagsAction` to enforce the rule that status can only get worse, not better. 
+
+`PropagatingTagsAction` will support multiple `String` values for a single tag, to support cases such as aggregating 
```

Comment:
Hey, I had that `NodeTagExtension` idea earlier. =) I'm not sure, honestly. Status is definitely a special case, but I don't know if it might not be more common. @i386, @michaelneale - any thoughts?

Custom data - this would be an advantage to the `List<String>` approach: we would at least guarantee that nothing more complex than a `String` could be stored, where `Object` is a little riskier, though we could still restrict it in practice to `String` or `List<String>`. I definitely don't want much here - either cases like status where the `String` itself is all the data involved, or the test result case where there'd be one or more `String`s but they'd just be pointers to more elaborate data on the `Run`...